lib/Dialect/FIRRTL/Transforms/LowerGroups.cpp

```diff
@@ -0,0 +1,375 @@
+              continue;
+            }
+
+      // For any other ops, create input ports for any captured operands.
+      for (auto operand : op.getOperands())
+        if (operand.getParentBlock() != body)
+          createInputPort(operand);
+    }
+
+    // Create the new module.  This grabs a lock to modify the circuit.
+    SmallString<32> newModuleName(moduleOp.getName());
+    newModuleName.append("_");
+    newModuleName.append(groupName);
+    FModuleOp newModule =
+        buildNewModule(builder, group.getLoc(), newModuleName, ports);
+    SymbolTable::setSymbolVisibility(newModule,
+                                     SymbolTable::Visibility::Private);
+    newModule.getBody().takeBody(group.getRegion());
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "      New Module: " << newModuleName << "\n";
+      llvm::dbgs() << "        ports:\n";
+      for (size_t i = 0, e = ports.size(); i != e; ++i) {
+        auto port = ports[i];
+        auto value = connectValues[i];
+        llvm::dbgs() << "          - name: " << port.getName() << "\n"
+                     << "            type: " << port.type << "\n"
+                     << "            direction: " << port.direction << "\n"
+                     << "            value: " << value << "\n";
+      }
+    });
+
+    // Replace the original group with an instance.  Hook up the instance.
+    builder.setInsertionPointAfter(group);
+    auto moduleName = newModule.getModuleName();
+    auto instanceOp = builder.create<InstanceOp>(
+        group.getLoc(), /*moduleName=*/newModule,
+        /*name=*/
+        (Twine((char)tolower(moduleName[0])) + moduleName.drop_front()).str(),
+        NameKindEnum::DroppableName,
+        /*annotations=*/ArrayRef<Attribute>{},
+        /*portAnnotations=*/ArrayRef<Attribute>{}, /*lowerToBind=*/true);
+    instanceOp->setAttr("output_file",
+                        hw::OutputFileAttr::getFromFilename(
+                            builder.getContext(),
+                            "groups_" + circuitName + "_" + groupName + ".sv",
+                            /*excludeFromFileList=*/true));
+    createdOps.insert(instanceOp);
+
+    // Connect instance ports to values.
+    assert(ports.size() == connectValues.size() &&
+           "the number of instance ports and values to connect to them must be "
+           "equal");
+    for (unsigned portNum = 0, e = newModule.getNumPorts(); portNum < e;
+         ++portNum) {
+      OpBuilder::InsertionGuard guard(builder);
+      builder.setInsertionPointAfterValue(instanceOp.getResult(portNum));
+      if (instanceOp.getPortDirection(portNum) == Direction::In)
+        builder.create<StrictConnectOp>(builder.getUnknownLoc(),
+                                        instanceOp.getResult(portNum),
+                                        connectValues[portNum]);
+      else
+        builder.create<StrictConnectOp>(
+            builder.getUnknownLoc(), connectValues[portNum],
```

Comment:
Not sure if can be confident `strictconnect` is safe re:uninferred widths (and passive?).
(and similarly for StrictConnect created above?)

EDIT: Oops, looks like we already restrict captures to be passive (?), so probably nevermind that bit :+1: .

Reply:
Yeah, this is super restricted in what it is allowed to be. This will cause problems as we start to open up this language feature.